### DISCUSSION TITLE
Pusher character object and step climbing interacting badly.

### DISCUSSION THREAD

**Discussion by Ono-Sendai** [jrouwe/JoltPhysics#399] [General]
Hi,
I'm using a CharacterVirtual for movement physics, and am adding an 'interaction' / 'pusher' object in order to push dynamic objects around as the player moves.

I'm following approach #2 from https://github.com/jrouwe/JoltPhysics/discussions/239, although I need to make the pusher object larger than the CharacterVirtual object for it to work (as opposed to smaller).

I'm following the "It follows the CharacterVirtual by setting the velocity" approach.

The issue I am running into seems to be a kind of feedback effect between step climbing a dynamic object and pushing the object: climbing a step seems to impart a reasonably large velocity, which in turn results in a large push on the step, which somehow causes the step to be climbed again in the next simulation step, or something like that.  The result is both the player and the object being climbed being flung with high velocity.

See https://www.youtube.com/watch?v=AFQIb9RVBXY for an example of this happening.

If I disable step climbing then it doesn't happen.

Any idea how to avoid this?  
I am actually smoothing out the the vertical step climbing position discontinuity but it's not helping.

**Comment by jrouwe** [comment]
From looking at the movie, I think the following is happening:

* The character virtual steps onto the dynamic object
* The rigid body spheres are moved to the new location
* The rigid body spheres intersect with the dynamic object and apply an impulse to the object
* The character virtual is standing on the object so it will assume the velocity of the dynamic object and move along
* The rigid body spheres are moved to the new location
* The rigid body spheres again intersect with the dynamic object and again apply an impulse to the object
* Etc. etc. etc.
* End result is that you fly away with the dynamic object

(the reason why it happens more with walk stairs is probably because it cheats a bit and gives a little bit more forward movement)

This is why it's important to have the rigid bodies that are attached to the character be smaller than the character. They should never intersect with anything the character is interacting with. The only reason they exist is to have other CharacterVirtual classes collide with the CharacterVirtual and to have high velocity bodies that use quality type LinearCast not teleport through the character virtual.

The character virtual pushes dynamic objects through other means, setting CharacterVirtualSettings::mMaxStrength should apply an impulse to anything the character interacts with. Put a breakpoint in CharacterVirtual::HandleContact and see why it doesn't get to the AddImpulse function call at the end of the function.

**Comment by Ono-Sendai** [comment]
Hi Jorrit, thanks for the detailed and helpful response.

I didn't realise that the CharacterVirtual can push around objects properly by itself.  I had noticed it pushing objects, but the force was so weak (my objects are quite heavy), that I assumed the force was just some kind of depenetration artifact.
Setting mMaxStrength to a larger value (such as 1e4) gives the kind of results I want.

Thanks!

**Comment by Ono-Sendai** [comment]
Actually this is still an issue even without any kind of pusher/interaction/rigid body.

See https://www.youtube.com/watch?v=8s9fwOZusco

Particularly for low mass objects that are stepped on it's a problem.

I wonder why stepping onto the object applies a force to it?

**Comment by jrouwe** [comment]
Hello,

What I see in the movie above is almost the same feedback loop as described earlier:

* The virtual character touches the object and applies a force to it
* The dynamic body responds by rotating
* Virtual character assumes the contact velocity of the rotating object, but as the player keeps pushing into the object it touches the object again and applies force to it again
* This happens until everything spins out of control (and a light object accelerates much more quickly, so the effect is much more visible)

The problem here is that the virtual character is not a regular dynamic object with inertia. It's a kinematic object so every frame when you set the velocity of the player you're adding energy to the system. The difficult thing with the character controller is that it is a gameplay construct that moves in a non-physical way, so that means that all interactions need to be 'designed' rather than that they follow from the laws of physics.

The reason why stepping onto an object applies a force is that this is the only way to e.g. make a boat rock if you step on it, or if you want the player to be able to interact with a seesaw (like in the CharacterVirtualDemo). You can turn this behavior off on a per object basis by overriding CharacterContactListener::OnContactAdded and setting ioSettings.mCanReceiveImpulses = false. You can for example base this on the contact normal: If the contact normal is approximately horizontal, allow interaction, if it is not, skip the impulses. The contact callback also receives the body ID so you can also take things like mass of the object into account when deciding if you want to interact. You can also set mCanPushCharacter = false if you don't want the object to push the player back.
